### PR TITLE
Update Distributed cache topic+sample

### LINKED ISSUE #8702: Distributed cache topic UE pass
.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 0d01025d-86b2-bb2d-86c1-ef0695bc98e3
* Version Independent ID: 5b9a6572-3671-0d54-b2ef-83d04f8b3e96
* Content: [Work with a distributed cache in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/performance/caching/distributed?view=aspnetcore-2.1)
* Content Source: [aspnetcore/performance/caching/distributed.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/performance/caching/distributed.md)
* Product: **aspnet-core**
* GitHub Login: @ardalis
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #8702 

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/distributed?view=aspnetcore-1.0&branch=pr-en-us-9129)

* The usual ...
  * Rotate the tires
  * Change the oil
  * Replace the pine tree air freshener
* Simplifying the title to make it easier to find in the TOC
* :tada: *Shiny new RP sample app* 🎈 ... simpler, too ... an old skool caching demo compared to what we had there before.
* Significant revision to the layout (sequence of sections/content)
* I want to add NCache to the topic ... *LATER* ... let's **NOT** do that right now. I'm going to open an issue for the work unless there's an objection. For now, I add links to NCache.

### CODE REVIEW

**Comment by pranavkm** [comment]
@sebastienros could you have a look?

File: aspnetcore/performance/caching/distributed.md [hunk 1]
```diff
@@ -1,146 +1,177 @@
...
+## Prerequisites
 
-**Refresh, RefreshAsync**
+::: moniker range=">= aspnetcore-2.1"
 
-Refreshes an item in the cache based on its key, resetting its sliding expiration timeout (if any).
+To use a SQL Server distributed cache, reference the [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app) or add a package reference to the [Microsoft.Extensions.Caching.SqlServer](https://www.nuget.org/packages/Microsoft.Extensions.Caching.SqlServer/) package.
 
-**Remove, RemoveAsync**
+To use a Redis distributed cache, reference the [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app) and add a package reference to the [Microsoft.Extensions.Caching.Redis](https://www.nuget.org/packages/Microsoft.Extensions.Caching.Redis/) package. The Redis package isn't included in the `Microsoft.AspNetCore.App` package, so you must reference the Redis package separately in your project file.
```
Comment (sebastienros) [inline aspnetcore/performance/caching/distributed.md]: There is a new package for 2.2, with a different id.

File: aspnetcore/performance/caching/distributed.md [hunk 2]
```diff
@@ -1,146 +1,177 @@
...
-> If using a SQL Server Distributed Cache, some of these advantages are only true if a separate database instance is used for the cache than for the app's source data.
-
-Like any cache, a distributed cache can dramatically improve an app's responsiveness, since typically data can be retrieved from the cache much faster than from a relational database (or web service).
-
-Cache configuration is implementation specific. This article describes how to configure both Redis and SQL Server distributed caches. Regardless of which implementation is selected, the app interacts with the cache using a common `IDistributedCache` interface.
+A distributed cache has the following advantages over other caching scenarios where cached data is stored on individual app servers:
 
-## The IDistributedCache Interface
+* Cached data is *coherent* (consistent) across requests to multiple servers.
+* Cached data survives server restarts and app deployments.
```
Comment (sebastienros) [inline aspnetcore/performance/caching/distributed.md]: Another reason to use it: cached data doesn't take local memory. (using an external cache)

File: aspnetcore/performance/caching/distributed.md [hunk 3]
```diff
@@ -1,146 +1,177 @@
...
-
-3. The source data store has fewer requests made to it (than with multiple in-memory caches or no cache at all).
-
-> [!NOTE]
-> If using a SQL Server Distributed Cache, some of these advantages are only true if a separate database instance is used for the cache than for the app's source data.
-
-Like any cache, a distributed cache can dramatically improve an app's responsiveness, since typically data can be retrieved from the cache much faster than from a relational database (or web service).
-
-Cache configuration is implementation specific. This article describes how to configure both Redis and SQL Server distributed caches. Regardless of which implementation is selected, the app interacts with the cache using a common `IDistributedCache` interface.
+A distributed cache has the following advantages over other caching scenarios where cached data is stored on individual app servers:
```
Comment (sebastienros) [inline aspnetcore/performance/caching/distributed.md]: These advantages only apply if not using the default implementation.

File: aspnetcore/performance/caching/distributed.md [hunk 4]
```diff
@@ -1,146 +1,177 @@
...
-The Redis cache requires [Microsoft.Extensions.Caching.Redis](https://www.nuget.org/packages/Microsoft.Extensions.Caching.Redis/)
+* [Distributed Memory Cache](#distributed-memory-cache)
+* [Distributed SQL Server cache](#distributed-sql-server-cache)
+* [Distributed Redis cache](#distributed-redis-cache)
 
-You configure the Redis implementation in `ConfigureServices` and access it in your app code by requesting an instance of `IDistributedCache` (see the code above).
+### Distributed Memory Cache
 
-In the sample code, a `RedisCache` implementation is used when the server is configured for a `Staging` environment. Thus the `ConfigureStagingServices` method configures the `RedisCache`:
+The Distributed Memory Cache (<xref:Microsoft.Extensions.DependencyInjection.MemoryCacheServiceCollectionExtensions.AddDistributedMemoryCache*>) is a framework-provided implementation of `IDistributedCache` that stores items in memory. The Distributed Memory Cache isn't an actual distributed cache. Cached items are stored by the app instance on the server where the app is running. The Distributed Memory Cache is a useful implementation in development and testing scenarios.
```
Comment (sebastienros) [inline aspnetcore/performance/caching/distributed.md]: I think it makes a lot of sense even in production when a single server is used and memory consumption is not an issue. It's a way to abstract the store, and allow for using a different implementation later on if multiple nodes are necessary.

File: aspnetcore/performance/caching/distributed.md [hunk 5]
```diff
@@ -1,146 +1,177 @@
...
+When deciding which implementation of `IDistributedCache` is best for your app, consider the following:
+
+* Existing infrastructure
+* Performance requirements
+* Cost
+* Team experience
+
+Caching solutions usually rely on in-memory storage to provide fast retrieval of cached data, but memory is a limited resource and costly to expand. Only store commonly used data in a cache.
+
+Generally, a Redis cache provides higher throughput and lower latency than a SQL Server cache. However, benchmarking is usually required to determine the performance characteristics of caching strategies.
```
Comment (sebastienros) [inline aspnetcore/performance/caching/distributed.md]: Also cache operation with SQL Server will impact the common database accesses if the same db is shared between the standard requests and caching.

**Comment by guardrex** [comment]
Thanks @sebastienros ... see how it looks on the last commit.

The only one I need to wait on is the Redis package for 2.2. I've noted it on the [2.2 RTM tracking issue (aspnet/Docs \#8518)](https://github.com/aspnet/Docs/issues/8518), which I'll get done the second 2.2 lands.

**Comment by guardrex** [comment]
@scottaddie Perhaps, we should go ahead with UE. I made all of Sébastien's changes ... except for the 2.2 Redis package. I don't see another package with a different id/version.

No rush on this btw.

File: aspnetcore/performance/caching/distributed/samples/2.x/DistCacheSample/SampleApp.csproj [hunk 6]
```diff
@@ -6,7 +6,6 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.App" />
```
Comment (scottaddie) [inline aspnetcore/performance/caching/distributed/samples/2.x/DistCacheSample/SampleApp.csproj]: Last I heard from @natemcmaster is we're recommending an explicit version now. Let's make that change here.

Comment (guardrex) [reply aspnetcore/performance/caching/distributed/samples/2.x/DistCacheSample/SampleApp.csproj]: Whaaaaaa ... again?

https://github.com/aspnet/Docs/issues/8579#issuecomment-424548306
https://github.com/dotnet/sdk/pull/2533

**Comment by guardrex** [comment]
I'm going to hold for a sec while I get more info on the double-reversal on `App` package version and to find out if I can put a script tag into this sample so it doesn't feel like it was built in 1995. *Yikes!* :smile: lol

File: aspnetcore/performance/caching/distributed/samples/2.x/DistCacheSample/SampleApp.csproj (hunk 6 above)
Comment (natemcmaster) [reply aspnetcore/performance/caching/distributed/samples/2.x/DistCacheSample/SampleApp.csproj]: @guardrex is correct. We're sticking with the implicit PackageRef version for 2.x. Although we temporarily removed our guidance about using implicit versions, we could not land on a satisfactory substitute for the 2.x timeframe. So, we restored guidance to use the versionless reference in https://github.com/aspnet/Docs/pull/8707.

FWIW we're refactoring all of this in 3.0 so there won't be a PackageReference to Microsoft.AspNetCore.App at all.

**Comment by guardrex** [comment]
ok ... the :pray: *server gods (Dan!)* :pray: has blessed us.

The styles will probably require some tweaking. We'll find out as we move forward dropping *wwwroot* and using this.

**Pull request merged**